File: src/electron/handlers/launcher.ts
import { ipcMain, app, BrowserWindow } from "electron";
import fs from "fs";
import path from "path";
import { spawn } from "child_process";

export function registerLauncherHandlers(win: BrowserWindow | null, __dirname: string) {
	ipcMain.handle("RunGame", async (_, optsJson, showLogs) => {
		try {
			const opts = typeof optsJson === "string" ? JSON.parse(optsJson) : optsJson;

			// Find instance binary
			let instanceBin = "";
			if (app.isPackaged) {
				instanceBin = path.join(process.resourcesPath, "goproton-instance");
			} else {
				instanceBin = path.resolve(__dirname, "../../bin/goproton-instance");
			}

			if (!fs.existsSync(instanceBin)) {
				console.error(`Instance binary not found at: ${instanceBin}`);
				instanceBin = "goproton-instance";
			}

			const args = [
				`--game=${opts.GamePath}`,
				`--launcher=${opts.LauncherPath || ""}`,
				`--prefix=${opts.PrefixPath}`,
				`--proton-pattern=${opts.ProtonPattern || ""}`,
				`--proton-path=${opts.ProtonPath || ""}`,
			];

			if (opts.EnableMangoHud) args.push("--mango");
			if (opts.EnableGamemode) args.push("--gamemode");
			if (opts.EnableGamescope) {
				args.push("--gamescope");
				if (opts.GamescopeW) args.push(`--gs-w=${opts.GamescopeW}`);
				if (opts.GamescopeH) args.push(`--gs-h=${opts.GamescopeH}`);
				if (opts.GamescopeR) args.push(`--gs-r=${opts.GamescopeR}`);
			}
			if (opts.EnableLsfgVk) {
				args.push("--lsfg");
				args.push(`--lsfg-mult=${opts.LsfgMultiplier || "2"}`);
				if (opts.LsfgPerfMode) args.push("--lsfg-perf");
				if (opts.LsfgDllPath) args.push(`--lsfg-dll-path=${opts.LsfgDllPath}`);
			}

			if (opts.EnableMemoryMin) {
				args.push("--memory-min");
				if (opts.MemoryMinValue) args.push(`--memory-min-value=${opts.MemoryMinValue}`);
			}

			// Pass logs flag explicitly
			if (showLogs === true) {
				args.push("--logs=true");
			} else if (showLogs === false) {
				args.push("--logs=false");
			}

			console.log("Spawning instance manager:", instanceBin, args.join(" "));

			const child = spawn(instanceBin, args, {
				detached: true,
				stdio: "ignore",
			});
			child.unref();

			app.quit();
			return null;
		} catch (e: any) {
			console.error("RunGame Error:", e);
			return e.toString();
		}
	});
}

File: src/instance/launcher.go
package main

import (
	"log"
	"strings"
)

// buildLaunchOptions creates the launch options from command line flags
func buildLaunchOptions() LaunchOptions {
	return LaunchOptions{
		GamePath:        gamePath,
		LauncherPath:    launcherPath,
		PrefixPath:      prefixPath,
		ProtonPattern:   protonPattern,
		ProtonPath:      protonPath,
		EnableMangoHud:  mango,
		EnableGamemode:  gamemode,
		EnableGamescope: gamescope,
		GamescopeW:      gsW,
		GamescopeH:      gsH,
		GamescopeR:      gsR,
		EnableLsfgVk:    lsfg,
		LsfgMultiplier:  lsfgMult,
		LsfgPerfMode:    lsfgPerf,
		LsfgDllPath:     lsfgDllPath,
		EnableMemoryMin: memoryMin,
		MemoryMinValue:  memoryMinValue,
	}
}

// logGameStartup logs the command and enabled features
func logGameStartup(cmdArgs []string) {
	log.Printf("--- EXECUTION START ---")
	log.Printf("COMMAND: %s", strings.Join(cmdArgs, " "))
	log.Printf("ENABLED FEATURES:")

	if mango {
		log.Printf("  [+] MangoHud")
	}
	if gamemode {
		log.Printf("  [+] GameMode")
	}
	if gamescope {
		log.Printf("  [+] Gamescope (%sx%s@%s)", gsW, gsH, gsR)
	}
	if lsfg {
		log.Printf("  [+] LSFG-VK (x%s, PerfMode:%v)", lsfgMult, lsfgPerf)
	}
	if memoryMin {
		log.Printf("  [+] Memory Protection (Min: %s)", memoryMinValue)
	}

	log.Printf("-----------------------")

	// Sync to disk so terminal sees output immediately
	if logFileHandle != nil {
		_ = logFileHandle.Sync()
	}
}
